tools/topology/m4/dai.m4

```diff
@@ -180,6 +180,7 @@ define(`DAI_ADD',
 `undefine(`SCHEDULE_DEADLINE')'
 `undefine(`SCHEDULE_PRIORITY')'
 `undefine(`SCHEDULE_CORE')'
+`undefine(`SCHEDULE_TIMER')'
```

Comment:
This would be a lot better if this was SCHEDULE_SOURCE where various SCHEDULE_SOURCE_s can be defined in schedule.m4. i.e.
```
# list of generic scheduling sources
define (SCHEDULE_SOURCE_DMA0, 0)
...
define (SCHEDULE_SOURCE_TIMER0, 100)
define (SCHEDULE_SOURCE_TIMER1, 101)
```
This would be then be made platform specific in apl.m4
```
define (SCHEDULE_SOURCE_HPGPDMA0, SCHEDULE_SOURCE_DMA0)
```
This then allows platform abstraction and the scheduler source value can be passed in as a tuple. This also means the platform scheduler source mappings must align between topology and FW.

Reply:
> This would be a lot better if this was SCHEDULE_SOURCE where various SCHEDULE_SOURCE_s can be defined in schedule.m4. i.e.
> 
> ```
> # list of generic scheduling sources
> define (SCHEDULE_SOURCE_DMA0, 0)
> ...
> define (SCHEDULE_SOURCE_TIMER0, 100)
> define (SCHEDULE_SOURCE_TIMER1, 101)
> ```
> This would be then be made platform specific in apl.m4
> 
> ```
> define (SCHEDULE_SOURCE_HPGPDMA0, SCHEDULE_SOURCE_DMA0)
> ```
> This then allows platform abstraction and the scheduler source value can be passed in as a tuple. This also means the platform scheduler source mappings must align between topology and FW.

um, I don't think it's a good idea as schedule source actually can be unlimited, it can be GPDMA or HDA-DMA(host/link), it could also be any possible components(e.g. volume/src/kpb/eq) if we extend our pipeline concept in the future(it can be host-less, dai-less, pipeline is only a special component(named bin) and can be connected to any other components/pipelines).

Reply:
Agreed, but that would require changes in pipeline ipc, and adding something like schedule_type/mode instead of current timer_delay. I wouldn't like to change the API now, because HDA pipelines are waiting to be fixed. Can we postpone this couple of weeks, until we agree on the best new interface approach including future needs.

Reply:
@keyonjie this is the whole point, the scheduling for each pipeline can have multiple configurations, but your change only allows one. We always have to think of the future when creating new APIs/ABIs.
@tlauda I can merge as long as this is converted to an ID rather than a flag in topology (that way we are not changing this API twice in weeks). We can then add the flexibility in a subsequent update with the ABI update.

Reply:
> @keyonjie this is the whole point, the scheduling for each pipeline can have multiple configurations, but your change only allows one. We always have to think of the future when creating new APIs/ABIs.
> @tlauda I can merge as long as this is converted to an ID rather than a flag in topology (that way we are not changing this API twice in weeks). We can then add the flexibility in a subsequent update with the ABI update.

@lgirdwood let's align on the intentions here as we had misunderstanding about what we wanted here:

sounds you are asking for something like "SCHEDULE_METHOD" configuration here, e.g.:
```
SCHEDULE_METHOD_INTERRUPT
SCHEDULE_METHOD_XTENSA_TIMER_0
SCHEDULE_METHOD_XTENSA_TIMER_1
SCHEDULE_METHOD_EXTERNAL_TIMER_0
```
Is that your point?

What we implemented here was actually adding something like "SCHEDULE_INTERVAL", such as:
```
pipeline 1(e.g. ssp0 playback): scheduled with DMA interrupt, SCHEDULE_INTERVAL for ppl1 set to be 0.
pipeline 2(e.g. ssp1 playback): scheduled with time interval(e.g. in every 1ms, not intend to specify and TIMER it will be used), SCHEDULE_INTERVAL for ppl2 set to be 1000.
pipeline 3(e.g. ssp2 playback): scheduled with time interval(e.g. in every 20ms, not intend to specify and TIMER it will be used), SCHEDULE_INTERVAL for ppl3 set to be 20000.
pipeline 4(e.g. HDMI1 playback): scheduled with time interval(e.g. in every 5ms, not intend to specify and TIMER it will be used), SCHEDULE_INTERVAL for ppl1 set to be 5000.
```

My point is that what kind of configurations the FW really needed from topologies, only SCHEDULE_INTERVAL, or SCHEDULE_METHOD + SCHEDULE_INTERVAL?

Reply:
> 
> sounds you are asking for something like "SCHEDULE_METHOD" configuration here, e.g.:
> 
> ```
> SCHEDULE_METHOD_INTERRUPT
> SCHEDULE_METHOD_XTENSA_TIMER_0
> SCHEDULE_METHOD_XTENSA_TIMER_1
> SCHEDULE_METHOD_EXTERNAL_TIMER_0
> ```
> Is that your point?

Yes. But it's better presented in English as SCHEDULE_INITIATOR (better than method or source)

> 
> What we implemented here was actually adding something like "SCHEDULE_INTERVAL", such as:
> 
> ```
> pipeline 1(e.g. ssp0 playback): scheduled with DMA interrupt, SCHEDULE_INTERVAL for ppl1 set to be 0.
> pipeline 2(e.g. ssp1 playback): scheduled with time interval(e.g. in every 1ms, not intend to specify and TIMER it will be used), SCHEDULE_INTERVAL for ppl2 set to be 1000.
> pipeline 3(e.g. ssp2 playback): scheduled with time interval(e.g. in every 20ms, not intend to specify and TIMER it will be used), SCHEDULE_INTERVAL for ppl3 set to be 20000.
> pipeline 4(e.g. HDMI1 playback): scheduled with time interval(e.g. in every 5ms, not intend to specify and TIMER it will be used), SCHEDULE_INTERVAL for ppl1 set to be 5000.
> ```

SCHEDULE_INTERVAL defines the time frequency/interval/period.
SCHEDULE_INITIATOR defines the time domain.

There can be many different time/clock domains within the audio architecture e.g. when SSP is slave then it uses a different time domain to another SSP which is master (even if MCLK, BCLK and FRAME are all the same frequency). We need to be able to configure this.

Reply:
In fact I would rename SCHEDULE_INTERVAL to SCHEDULE_PERIOD or SCHEDULE_FREQ as these make more sense than interval.

Reply:
OK, let me rename this SCHEDULE_TIMER to SCHEDULE_PERIOD, and don't touch the SCHEDULE_INITIATOR at this moment, does that make sense?

Reply:
@keyonjie not following, wont you need to specify the initiator to make HDA use timer ?

Reply:
> @keyonjie not following, wont you need to specify the initiator to make HDA use timer ?

no, we didn't intend to specify the initiator, we only wanted to specify the SCHEDULE_PERIOD actually.